format docstrings with CommonMark

bramtayl opened the pull request.

bramtayl commented:
Rebase #254

bramtayl commented:
I should also mention while doing this this that I realized style and state don't propagate to `julia>` doctests.... When I tried to do that, it messed up the indentation.

bramtayl commented:
Also, this won't work until the next version of CommonMark is tagged containing MichaelHatherly/CommonMark.jl#5

bramtayl commented:
Ok...updated it to propagate state and style. There' still something going on weird with indentation though...

StefanKarpinski commented on test/files/Onda.jl/src/samples.jl:
```diff
@@ -108,10 +107,10 @@ channel(samples::Samples, i::Integer) = channel(samples.signal, i)
-    the former is the duration of the entire original signal in the context of its
-    parent recording, whereas the latter is the actual duration of `samples.data`
-    given `samples.signal.sample_rate`.
+`duration(samples)` is not generally equivalent to `duration(samples.signal)`;
```
Seems not to understand note syntax.

MichaelHatherly replied:
You'll need `[AdmonitionRule(), FootnoteRule(), MathRule(), TableRule()]`. I *think* that's everything needed, but their may be a few other small differences such as https://github.com/MichaelHatherly/CommonMark.jl/issues/6, which may be a deal breaker for this until it's resolved.

bramtayl replied:
Cool! Thanks for the help. I don't see that issue as a big problem as long as it'll be fixed eventually.

domluna commented on src/pretty.jl:
```diff
@@ -227,7 +227,77 @@ end
+                indended_state = State(
+                    state.doc,
+                    state.indent_size,
+                    state.indent + 4,
```
i think the code blocks are indented so that may be why. If that's the case it should be `state.indent + state.indent_size`

bramtayl replied:
Still not sure I understand...using `indent_size` seems to work though...Is there an easy way to shift all lines in each input chunk over by seven spaces (to account for the `julia> `)?

bramtayl commented:
Ok, I've added most of the other extensions. @MichaelHatherly I'm still hitting bugs when I enable MathRule in the PR, even after https://github.com/MichaelHatherly/CommonMark.jl/commit/48c2ecc9e3e209af04c53453baf59d3636f7a245. I still can't figure out what's going on with the indenting...

MichaelHatherly commented:
> Ok, I've added most of the other extensions. @MichaelHatherly I'm still hitting bugs when I enable MathRule or TableRule in the PR, even after [MichaelHatherly/CommonMark.jl@48c2ecc](https://github.com/MichaelHatherly/CommonMark.jl/commit/48c2ecc9e3e209af04c53453baf59d3636f7a245). I still can't figure out what's going on with the indenting...

Even if they aren't minimally reproducible could you open an issue with what's not working correctly and I'll attempt to track it down, thanks.

bramtayl commented:
Ok, will do. I was having trouble triggering them without running the full test suite, but I'll give it a shot.

bramtayl commented:
Ok, so updates:

Thanks to the help of @MichaelHatherly, I've been able to enable all the CommonMark plugins

I figured out part of the indenting problem. I had to indent all lines in the docstring to match the indent of the first line...but now I added another bug? Anyway, there's an extra `@test_broken` that I don't know how to fix.

domluna commented on test/default_style.jl:
```diff
@@ -4795,4 +4803,129 @@ some_function(
+        function test() end"""
+        @test format_text(unformatted) == formatted
+    end
+    @testset "Indented docstring" begin
+        unformatted = """
+        begin
+            "Indented docstring"
+            indented_item
+        end"""
+        formatted = """
+        begin
+            \"""
+            Indented docstring
+            \"""
+            indented_item
+        end"""
+        @test format_text(unformatted) == formatted
+        @test_broken format_text(formatted) == formatted
```
What's the output difference for this?

bramtayl replied:
It ends up being

```
"""
begin
    \"""
        Indented docstring
    \"""
    indented_item
end
"""

```

domluna replied:
the code blocks are indented so maybe It thinks it's that? 

btw you should use fmt instead of format_text for testing because `fmt` calls format_text twice so it ensures first format == second format

bramtayl commented:
Ok..., well, I've finally sorted out the indenting issues. I heavily commented the code where I fixed docstring indenting to handle the `"""`. I've also updated format to format markdown files in the main folder as if they were docstrings.

bramtayl commented:
I don't know why this is getting flagged as not formatted, because I formatted the code locally...

domluna commented:
@bramtayl if you run `format(".")` from the root dir it should be good

bramtayl commented:
Lots of green checkmarks!

bramtayl commented:
I'm just going to add a test for formatting plain markdown documents and then I think it will be ready

bramtayl commented:
Ok, I'm happy

domluna commented:
:clap: NICE!!! :clap: I'll have a final look tonight or tmrw

domluna commented on test/files/Onda.jl/LICENSE.md:
```diff
@@ -1,22 +1,21 @@
 
 > Copyright (c) 2019: Beacon Biosignals, Inc.
->
+> 
```
is this whitespace? kinda weird that's added right?

bramtayl replied:
Yes, it is. I could open up an issue in CommonMark?

domluna commented on test/files/setup_databases.jl:
```diff
@@ -35,24 +37,32 @@ function parse_commandline()
         help = "path where the database folder is going to be created."
         default = "."
-        "--pdb"
+        """
+        --pdb
+        """
```
I personally don't mind this and I don't think there's an easy way to get around this.

bramtayl replied:
I thought about changing it to do oneliners with single quotes, but I think I like this better. It makes it easier to add more to the docstring later on

bramtayl commented:
I'm happy with it, and I think it would be great as an option. Sorry about the messy commit history
